PackageKit/backends/zypp/pk-backend-zypp.cpp

```diff
@@ -182,6 +184,73 @@ gchar * _repoName;
  */
 gboolean _updating_self = FALSE;
 
+// Forward declaration
+static void
+zypp_backend_finished_error (PkBackendJob  *job, PkErrorEnum err_code,
+			     const char *format, ...);
+
+static gboolean
+zypp_set_dist_upgrade_mode (gboolean dist_upgrade_mode)
+{
+	const char *target_path = "/var/cache/pk-zypp-cache";
+	const char *path = dist_upgrade_mode ? "/var/cache/pk-zypp-dist-upgrade"
+		: "/var/cache/zypp";
+
+	PK_ZYPP_LOG ("%s: %s -> %s", __func__, target_path, path);
+
+	char tmp[PATH_MAX];
+	struct stat st;
+
+	if (stat(path, &st) != 0) {
+		PK_ZYPP_LOG ("Creating cache directory: %s", path);
+		if (mkdir(path, 0755) != 0) {
+			PK_ZYPP_LOG ("Cannot create directory: %s", strerror(errno));
+			return FALSE;
+		}
+	}
+
+	if (lstat(target_path, &st) != 0) {
+		PK_ZYPP_LOG ("Creating symlink: %s -> %s", target_path, path);
+		if (symlink(path, target_path) != 0) {
+			PK_ZYPP_LOG ("Cannot create symlink: %s", strerror(errno));
+			return FALSE;
+		}
+		return TRUE;
+	}
+
+	if (readlink(target_path, tmp, sizeof(tmp)) == -1) {
+		PK_ZYPP_LOG ("Cannot read dist upgrade path: %s", strerror(errno));
+		return FALSE;
+	}
+
+	// If target_path is already pointing to path, we're done
+	if (strcmp(path, tmp) == 0) {
```

Comment:
This strcmp compare with an uncorrectly terminated string, which is why the symlink gets re-created too often, even if the path has not changed.
See what debug output placed before this line produces:

   Comparing path /var/cache/zypp with tmp /var/cache/zyppche/zypp